### ISSUE TITLE
Can I use coveralls from a single repo with two different code bases?

### ISSUE THREAD

**Issue by dominikwilkowski** [lemurheavy/coveralls-public#1640]
I have a repo that includes a node and a rust implementation of the same thing. I have coveralls setup for the node implementation but now I don't know how to add the rust one. I have a `lcov.info` file for the rust coverage but since it comes from the same repo I fear it would just override the node one...

Is there a way to create multiple coveralls project for the same repo basically?

**Comment by afinetooth** [comment]
Hi, @dominikwilkowski.

While there's no way to create multiple coveralls projects for a single github repo (the relationship is 1-to-1), yours is a common case and this is the recommended approach:

You will probably want to configure coveralls.io to receive **parallel builds** for your project. 

Specifically, what you'll do is modify your CI config file to build your project in parts, in parallel, often called a matrix build, where you'll send the node project coverage to coveralls.io in _one_ parallel "job" and send the rust project coverage in _another_ parallel job.

You're ahead of the game with a `lcov.info` file for your rust implementation. Nearly all coveralls integrations accept and process `LCOV` format coverage reports.

One integration that will be particularly useful / efficient for your use case is the [Coveralls Github Action](https://github.com/marketplace/actions/coveralls-github-action), which _requires_ `LCOV` reports. 

[This usage example from the README](https://github.com/marketplace/actions/coveralls-github-action#complete-parallel-job-example) demonstrates how to configure your Github Actions config YAML to send parallel jobs to coveralls.io.

In addition, the [coveralls-node-demo](https://github.com/coverallsapp/coveralls-node-demo) project demonstrates the approach to configuring parallel builds for coveralls.io in the more general case, providing examples for Github Actions, CircleCI and Travis.

A so-called "matrix build" might not make sense for your project as they typically apply to running the same codebase against multiple different environments (like OS's). 

Instead, you might just want to set up your project's CI builds with two (2) parallel jobs, in three (3) steps, like this:

### Example using Coveralls Github Action

1. **Step 1**: Run the tests for the node portion of your project. Convert the coverage report to `LCOV` format if necessary. Send the coverage report to coveralls with:

```
jobs:
  tests:
    steps:
      name: Coveralls Node Parallel
      uses: coverallsapp/github-action@master
      with:
        github-token: ${{ secrets.github_token }}
        parallel: true
        path-to-lcov: <path to `lcov.info` if not at `./coverage/lcov.info`>
        flag-name: node-job
```

2. **Step 2**: Run the tests for the rust portion of your project. Convert the coverage report to `LCOV` format if necessary (sounds like it's not necessary). Send the coverage report to coveralls with:

```
jobs:
  tests:
    steps:
      [...]
  
      name: Coveralls Rust Parallel
      uses: coverallsapp/github-action@master
      with:
        github-token: ${{ secrets.github_token }}
        parallel: true
        path-to-lcov: <path to `lcov.info` if not at `./coverage/lcov.info`>
        flag-name: rust-job
```

3. **Step 3**: Close the parallel build by sending the [parallel build (closed) webhook](https://docs.coveralls.io/parallel-build-webhook), like this:

```
  finish:
    needs: tests
    steps:
      - name: Coveralls Parallel Build Finished
        uses: coverallsapp/github-action@v1.1.0
        with:
          github-token: ${{ secrets.github_token }}
          parallel-finished: true
```

The third step here is required. You must tell coveralls.io to _close_ the parallel build before it will merge coverage reports from each job and calculate an aggregate coverage % for the full project.

Then you'll have both a single coverage report for the project as a whole, as well as individual coverage reports for each of the constituent jobs.

Hope that helps!

**Comment by dominikwilkowski** [comment]
Thank you for the detailed answer @afinetooth.

This has worked really well and for others you can see the PR here: https://github.com/dominikwilkowski/cfonts/pull/51

Note some things I had to do with `base-path` but works like a charm now :)

**Comment by dominikwilkowski** [comment]
Actually I'm confused now. Why is it now showing only rust?
https://coveralls.io/github/dominikwilkowski/cfonts

The pull request (https://coveralls.io/builds/49785114) itself seems to show the both languages but once merged it now only shows rust or am I reading this wrong somehow?

**Comment by afinetooth** [comment]
Hi @dominikwilkowski.

So it looks like you're only configured for parallel builds on **pull_requests**. For example, for this PR build:
https://coveralls.io/builds/49785114

You can see your two parallel jobs:
![Screen Shot 2022-06-07 at 9 48 16 AM](https://user-images.githubusercontent.com/918667/172438132-c583daae-5a6e-41c3-8145-20ea98ab12b7.png)

And your SOURCE FILES table shows both rust and JS files:
![Screen Shot 2022-06-07 at 9 48 46 AM](https://user-images.githubusercontent.com/918667/172438232-5047390e-802d-4513-942d-d58b6a96ab77.png)

On the other hand, your **push** type builds on your _default_ branch (**released**), still appear to be non-parallel. Example:
https://coveralls.io/builds/49785123

Only one job:
![Screen Shot 2022-06-07 at 9 50 05 AM](https://user-images.githubusercontent.com/918667/172438449-02e39c05-6a86-4a34-ba63-b005405ccd51.png)

Only rust files:
![Screen Shot 2022-06-07 at 9 50 32 AM](https://user-images.githubusercontent.com/918667/172438534-413f6150-f7b9-4db9-8f22-6f491094e3f5.png)

The reason you're seeing the same thing (only rust files in your SOURCE FILES table) on your main repo page:
https://coveralls.io/github/dominikwilkowski/cfonts

Is because the repo page always reflects the _last_ build on your _default_ branch (**released**).

That's because it is meant to reflect the coverage % for the production (or production-ready) version of your code. Same as your badge coverage is meant to represent.

If you will share your config file, we can see why your **push** builds aren't building in parallel.

My guess, right now, since the one job in your **push** builds reads, `.2`:
https://coveralls.io/builds/49785123

![Screen Shot 2022-06-07 at 9 53 18 AM](https://user-images.githubusercontent.com/918667/172439037-b14fde7b-7516-48b3-8e6d-d73b633ba929.png)

Is that `parallel: true` is missing for one step and the second job (rust) is replacing the first job (JS).

Have a look at the usage example for parallel builds with the Coveralls Github Action again:
https://github.com/marketplace/actions/coveralls-github-action#complete-parallel-job-example

Note how the `on:` key specifies the same build behavior for both `push` and `pull_request` type builds:

```
on: ["push", "pull_request"]

name: Test Coveralls Parallel
[...]
```

**Comment by dominikwilkowski** [comment]
You can see my config here: https://github.com/dominikwilkowski/cfonts/blob/released/.github/workflows/testing.yml

```yml
on:
  push:
    branches:
      - 'released'
    tags:
      - '!v*'
  pull_request:
    branches:
      - 'released'
```

And I have `parallel: true` on both jobs. The only difference is that I have to specify `base-path: ./nodejs` on one of them presumably because the `lcov.info` paths perhaps?

(for the chance it changes later here the relevant bits)
```yml
# -------------------------------| COVERAGE |------------------------------- #
  # ***********
  # COVERAGE - nodejs
  # ***********
  coverage-nodejs:
    needs: test-nodejs
    strategy:
      matrix:
        node: [16]
        os:
          - ubuntu-latest
    runs-on: ${{ matrix.os }}
    defaults:
      run:
        working-directory: ./nodejs

    steps:
      - uses: actions/checkout@v2

      - name: Setup Node
        uses: actions/setup-node@v3
        with:
          node-version: ${{ matrix.node }}

      - name: Get yarn cache
        id: yarn-cache
        run: echo "::set-output name=dir::$(yarn cache dir)"

      - name: Cache dependencies
        uses: actions/cache@v2
        with:
          path: ${{ steps.yarn-cache.outputs.dir }}
          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-yarn-
      - name: Yarn install dependencies
        run: yarn install --frozen-lockfile

      - name: Build files
        run: yarn build

      - name: Produce Coverage
        run: yarn jest --coverage

      - name: Print last few lines of lcov.info
        run: tail ./coverage/lcov.info

      - name: Upload to coveralls
        uses: coverallsapp/github-action@master
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          path-to-lcov: ./nodejs/coverage/lcov.info
          parallel: true
          base-path: ./nodejs

  # ***********
  # COVERAGE - rust
  # ***********
  coverage-rust:
    needs: test-rust
    strategy:
      matrix:
        os:
          - ubuntu-latest
    runs-on: ${{ matrix.os }}
    defaults:
      run:
        working-directory: ./rust

    steps:
      - uses: actions/checkout@v2

      - uses: actions-rs/toolchain@v1
        with:
          toolchain: stable

      - name: Run Makefile
        run: make

      - name: Install Rust
        run: rustup toolchain install stable --component llvm-tools-preview

      - name: Install cargo-llvm-cov
        uses: taiki-e/install-action@cargo-llvm-cov

      - name: Generate code coverage
        run: cargo llvm-cov --all-features --workspace --lcov --output-path lcov.info

      - name: Print last few lines of lcov.info
        run: tail ./lcov.info

      - name: Upload to coveralls
        uses: coverallsapp/github-action@master
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          path-to-lcov: ./rust/lcov.info
          parallel: true

  # ***********
  # COVERAGE - finalizing
  # ***********
  coverage-finish:
    needs: [coverage-nodejs, coverage-rust]
    strategy:
      matrix:
        os:
          - ubuntu-latest
    runs-on: ${{ matrix.os }}
    steps:
      - name: Coveralls Parallel Build Finished
        uses: coverallsapp/github-action@v1.1.0
        with:
          github-token: ${{ secrets.github_token }}
          parallel-finished: true
```

**Comment by afinetooth** [comment]
Hi @dominikwilkowski. Ok, yes, I don't see anything in your config that indicates why we'd only be getting non-parallel builds from **released**. 

I cover some observed details on that below, but the important thing is the **next step**, which is to enable [verbose mode](https://docs.coveralls.io/troubleshooting#put-it-in-verbose--v) on your coveralls integration so we can search for clues in the debug info that's added to your CI build logs.

---

**Next step**:
Enable verbose mode for the [Coveralls Github Action](https://github.com/marketplace/actions/coveralls-github-action):

- Add the following environment variable to your GHA config YAML (or secrets) so that it's available to any/all coveralls steps: `NODE_COVERALLS_DEBUG=1` (The `NODE_` part is there because the [Coveralls Github Action](https://github.com/marketplace/actions/coveralls-github-action) uses the [node-coveralls](https://github.com/nickmerwin/node-coveralls) integration under-the-hood.)

---

**Details from CI build logs**:
Relevant, if not conclusive:

- Your [most recent push build on released](https://coveralls.io/builds/49785123) points to [this log](https://github.com/dominikwilkowski/cfonts/actions/runs/2452417481), which shows both coverage jobs were sent to coveralls.io:
![Screen Shot 2022-06-08 at 10 20 36 AM](https://user-images.githubusercontent.com/918667/172678399-ff9b5ee1-016a-4e5e-9ae2-13d2ab2aa528.png)

Neither job shows any failures, but the second job appears to have replaced the first:
https://coveralls.io/builds/49785123

![Screen Shot 2022-06-08 at 10 21 31 AM](https://user-images.githubusercontent.com/918667/172678574-db85b742-ac60-4be0-a61f-7d0833f5b977.png)

_Again, verbose / debug info should help with clues why._

**Comment by dominikwilkowski** [comment]
I enabled the `NODE_COVERALLS_DEBUG` env var and re-ran the action here: https://github.com/dominikwilkowski/cfonts/actions/runs/2452417481

Also just in case the repeat of the action doesn't actually take the new env var into account I created a new PR here: https://github.com/dominikwilkowski/cfonts/pull/52

And the PR shows again both builds: https://coveralls.io/builds/49859798
I assume it will drop back to only rust once merged just like before since we didn't change anything material.

**Comment by afinetooth** [comment]
Thanks, @dominikwilkowski.

I'm not seeing the debug output I'd expect from you adding the `NODE_COVERALLS_DEBUG=1` env var. I could not find `NODE_COVERALLS_DEBUG=1` in the [`.github/workflows/testing.yml` file for that commit](https://github.com/dominikwilkowski/cfonts/blob/ea77f075f9c1138bf753256d1a82c027cb2d0d43/.github/workflows/testing.yml), so maybe you added it as a repo secret?

In any case, to benefit from that debug info, it would be great if you could try adding that again. It should be as easy as adding and env: key to each coveralls-related step, something like:

```
name: Upload to coveralls
        uses: coverallsapp/github-action@master
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          path-to-lcov: ./nodejs/coverage/lcov.info
          parallel: true
          base-path: ./nodejs
        env:
          NODE_COVERALLS_DEBUG=1
```

Aside from that, like you do I more or less expect to see the same dropping of one job in the next push build on released. If only because, at this point, we haven't really changed anything.

But I do believe there will be clues in the verbose logs for both the next PR build and the next push build.

**Comment by dominikwilkowski** [comment]
Hey @afinetooth 

I just added it to the testing yaml file as well. I did add it as a secret.
Hope this works now

**Comment by afinetooth** [comment]
HI @dominikwilkowski. Me too!. Let me know.

**Comment by dominikwilkowski** [comment]
@afinetooth 
So do you want me to merge it? I will do that now and see. I can always open a new PR up for debugging.

Thank you btw for your excellent help. It's been a great experience.

**Comment by dominikwilkowski** [comment]
I actually merged another PR yesterday https://github.com/dominikwilkowski/cfonts/pull/53 that also resulted in the right view on the PR in coveralls so it showed both rust and nodejs (https://coveralls.io/builds/49913169) but upon merging now only shows rust alone (https://coveralls.io/github/dominikwilkowski/cfonts). Any help is appreciated.

**Comment by dominikwilkowski** [comment]
Ok that PR is now also merged (interesting that the finishing job doesn't really generate verbose output like the other once https://github.com/dominikwilkowski/cfonts/runs/6840306155?check_suite_focus=true) and coveralls still only shows rust as a language.

**Comment by afinetooth** [comment]
Hi, @dominikwilkowski.

Thanks for the updates. 

This behavior is disconcerting since I don't see any reason for it in your config file.

That's why I need to dig deeper into your verbose verbose CI build logs.

However, using the last **[PR build](https://coveralls.io/builds/49913169)** and **[base build](https://coveralls.io/builds/49785123)** pairing you mention as examples, I'm not seeing verbose output in those build logs:

### Builds

**PR build**:
https://coveralls.io/builds/49913169

<img width="1141" alt="Screen Shot 2022-06-13 at 11 14 11 AM" src="https://user-images.githubusercontent.com/918667/173417999-2ad41bd4-461c-46c7-bef7-e323f5630116.png">

**Base build**: 
https://coveralls.io/builds/49785123

<img width="1147" alt="Screen Shot 2022-06-13 at 11 14 35 AM" src="https://user-images.githubusercontent.com/918667/173418063-3b5dcef7-9f4c-4228-a2b7-6d2c05d122a3.png">

### Logs

**[PR build](https://coveralls.io/builds/49913169)** log (Job 1):
https://github.com/dominikwilkowski/cfonts/runs/6826526223?check_suite_focus=true

<img width="1552" alt="Screen Shot 2022-06-13 at 11 16 09 AM" src="https://user-images.githubusercontent.com/918667/173418356-32e53f75-dfaf-4108-8ad5-26d0a262adc7.png">

**[PR build](https://coveralls.io/builds/49913169)** log (Job 2):
https://github.com/dominikwilkowski/cfonts/runs/6826566931?check_suite_focus=true

<img width="1552" alt="Screen Shot 2022-06-13 at 11 16 38 AM" src="https://user-images.githubusercontent.com/918667/173418430-7fbd279c-6c83-47df-8a7a-b14c4d315a39.png">

And:

**[Base build](https://coveralls.io/builds/49785123)** log (First job - missing from Coveralls build page): 
https://github.com/dominikwilkowski/cfonts/runs/6801302087?check_suite_focus=true

<img width="1552" alt="Screen Shot 2022-06-13 at 11 18 48 AM" src="https://user-images.githubusercontent.com/918667/173418802-62f7b13a-bff9-4f8c-b074-0c73a5dfd827.png">

**[Base build](https://coveralls.io/builds/49785123)** log (Second job - Appears in Coveralls build page as Job `.4`): 
https://github.com/dominikwilkowski/cfonts/runs/6801317417?check_suite_focus=true

<img width="1552" alt="Screen Shot 2022-06-13 at 11 19 47 AM" src="https://user-images.githubusercontent.com/918667/173418953-808afd14-8a9a-4582-b6d1-74184e838701.png">

_BTW, you're right, the "Coveralls Finish" job (aka. "Parallel build closed webhook" job):
https://github.com/dominikwilkowski/cfonts/runs/6801345192?check_suite_focus=true_

_Does not currently have any helpful debug output for this integration_:

<img width="1552" alt="Screen Shot 2022-06-13 at 11 23 37 AM" src="https://user-images.githubusercontent.com/918667/173419583-9ccbab51-16b3-4116-8041-67135d25dcb3.png"> 

_I'm hoping we'll add some soon, but for the success case, there's not all that much to see beyond the BUILD ID in the original request, and, unless there's a failure, all that happens in the success case is that the build's status = `done`, meaning it's closed at coveralls.io, which we can easily confirm in the web UI. That's how I use it._ 

_The meat of the sandwich is in the two coverage job POSTs, particularly the JSON object representing the contents of the JSON file sent as a form file upload with each POST._

ANYWAY:
I'm not seeing any debug output in the CI logs, as you can see from the screenshots above), and I'm not finding the `NODE_COVERALLS_DEBUG=1` env var in your [config file](https://github.com/dominikwilkowski/cfonts/blob/ea77f075f9c1138bf753256d1a82c027cb2d0d43/.github/workflows/testing.yml):

<img width="1552" alt="Screen Shot 2022-06-13 at 11 28 12 AM" src="https://user-images.githubusercontent.com/918667/173420364-ba735728-5700-4862-865e-b98a1c83151b.png">

So, questions:

1. Did you add `NODE_COVERALLS_DEBUG=1` as a secret instead?
2. Are you seeing debug output somewhere? Am I just missing it?

If you did run some verbose builds, then destroy / replace them due to security concerns with having them up at Github, can you re-run the builds, export the verbose logs and share those with me at support@coveralls.io? If you mention [this issue](https://github.com/lemurheavy/coveralls-public/issues/1640), I'll get them.

---

**Comment by dominikwilkowski** [comment]
Hey,
Let's pick the last PR that I have merged. I'm also happy to open a new one for you to investigate this. Do let me know if that would help.


Last PR: https://github.com/dominikwilkowski/cfonts/pull/55
Actions for it: https://github.com/dominikwilkowski/cfonts/actions/runs/2479518322
  - Action for nodejs parallel job: https://github.com/dominikwilkowski/cfonts/runs/6842439021?check_suite_focus=true
  - Action for rust parallel job: https://github.com/dominikwilkowski/cfonts/runs/6842444372?check_suite_focus=true
  - Action to finalize: https://github.com/dominikwilkowski/cfonts/runs/6842454552?check_suite_focus=true
Debug flags: https://github.com/dominikwilkowski/cfonts/blob/released/.github/workflows/testing.yml#L224
Coveralls build on this branch: https://coveralls.io/builds/49941830

I hope this helps you look into this

To your questions:

>     Did you add NODE_COVERALLS_DEBUG=1 as a secret instead?

I ended up doing both. The current workflow actually includes them on the default branch here: https://github.com/dominikwilkowski/cfonts/blob/released/.github/workflows/testing.yml

>     Are you seeing debug output somewhere? Am I just missing it?

Yes I am seeing the debug output for the parallel builds. Screenshots below (for the links above)

Nodejs:
<img width="1276" alt="image" src="https://user-images.githubusercontent.com/1266923/173450399-49c5af67-8fb1-41a1-93fe-9727cb2d42e4.png">

Rust:
<img width="1276" alt="image" src="https://user-images.githubusercontent.com/1266923/173450535-902796ca-3b73-44f2-8a49-533f8168b815.png">

**Comment by afinetooth** [comment]
Ok, that's great. That has everything I need.

So, as discussed, the PR builds are acting normally. Everything I see in those logs is expected. And the open question is why the **push** builds aren't acting the same.

In this context, the **base build** / **push build** is this one:
https://coveralls.io/builds/49940082

And luckily we have verbose logs for both of its jobs, here (**nodejs**):
https://github.com/dominikwilkowski/cfonts/runs/6842439021?check_suite_focus=true

![Screen Shot 2022-06-14 at 11 36 35 AM](https://user-images.githubusercontent.com/918667/173664002-cd8650fc-5e2d-4420-9df9-e20f77eb912c.png)

And here (**rust**):
https://github.com/dominikwilkowski/cfonts/runs/6842444372?check_suite_focus=true

![Screen Shot 2022-06-14 at 11 36 55 AM](https://user-images.githubusercontent.com/918667/173664072-e2fd143d-dd04-4e75-9737-76511803a867.png)

Notice that both show `200 success` messages, and display the resulting **new job URL**.

The difference is that the first job (**nodejs**) is missing from coveralls.io:
https://coveralls.io/jobs/101054720

![Screen Shot 2022-06-14 at 11 33 22 AM](https://user-images.githubusercontent.com/918667/173663424-99c55fec-bc56-4061-8b17-50f7ba41cb86.png)

While the second job (**rust**) is there:
https://coveralls.io/jobs/101054767

![Screen Shot 2022-06-14 at 11 34 33 AM](https://user-images.githubusercontent.com/918667/173663612-cbc5a994-8458-4ca3-8674-eea756b468b9.png)

---

So it's as though the second job (**rust**) replaced the first (**nodejs**).

For which there would be two (2) possible reasons that I'm aware of:

1. The build doesn't know it's _parallel_
2. Both jobs have identical information, making the second job appear to be a _duplicate_ of the first job.

But I don't see evidence for _either_ of those two reasons in the verbose build logs.

So, unfortunately, I need to keep searching for some other, unexpected reason, which I'll do and get back here, asap.

**Comment by afinetooth** [comment]
@dominikwilkowski I think I've got it.

Please try adding a `flag-name` parameter to the config for each `coveralls/upload` job, per the instructions here:
https://github.com/marketplace/actions/coveralls-github-action#inputs

While, under **Inputs** section there, the `flag-name` option is listed as _optional_, it's actually _required_ in the case of a **parallel job** (`parallel: true`):

> **parallel** 
> _optional_ 
> Set to `true` for parallel (or matrix) based steps, where multiple posts to Coveralls will be performed in the check. `flag-name` _needs to be **set** and **unique**_, e.g. `flag-name: run-${{ matrix.test_number }}`

Visually:

![Screen Shot 2022-06-14 at 11 44 54 AM](https://user-images.githubusercontent.com/918667/173666740-e0a35846-e6f8-47be-bb49-7af0ef813461.png)

Since your jobs are not in a matrix, you don't have to use a variable in the name, just use any name that's unique between the two jobs.

Let us know how it goes.

**Comment by dominikwilkowski** [comment]
This worked! yay!
Thank you so much for all your help. amazing experience.

**Comment by afinetooth** [comment]
You're welcome!
